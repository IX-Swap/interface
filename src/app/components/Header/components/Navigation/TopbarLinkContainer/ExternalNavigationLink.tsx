import { Link, Typography } from '@mui/material'
import React from 'react'
import { useStyles } from 'app/components/Header/components/Navigation/TopbarLinkContainer/TopbarLinkContainer.styles'
import { TopbarLinkProps } from './TopbarLinkContainer'

export const ExternalNavigationLink = ({
  link,
  label,
  placement
}: TopbarLinkProps) => {
  const classes = useStyles({
    isActive: false,
    disabled: false,
    placement
  })
  return (
    <Link href={link} className={classes.wrapper} target={'_blank'}>
      <Typography className={classes.text} variant={'body1'}>
        {label}
      </Typography>
    </Link>
  )
}
