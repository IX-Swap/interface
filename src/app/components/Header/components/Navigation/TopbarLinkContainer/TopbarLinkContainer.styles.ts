import makeStyles from '@mui/styles/makeStyles'

export interface Props {
  isActive: boolean
  disabled: boolean
  placement?: 'topbar' | 'dropdown' | 'mobileDropdown'
}

export const useStyles = makeStyles(theme => ({
  wrapper: {
    display: 'flex',
    alignItems: 'center',
    paddingLeft: theme.spacing(2.5),
    paddingRight: theme.spacing(2.5),
    paddingTop: theme.spacing(2),
    paddingBottom: theme.spacing(2),
    width: (props: Props) => (props.disabled ? 'auto' : '100%'),
    height: (props: Props) => (props.placement === 'topbar' ? 80 : 'initial'),
    boxSizing: 'border-box',
    fill: (props: Props) =>
      props.isActive
        ? theme.palette.navigationLink.activeColor
        : theme.palette.navigationLink.color,
    color: (props: Props) =>
      props.isActive
        ? theme.palette.navigationLink.activeColor
        : theme.palette.navigationLink.color,
    whiteSpace: 'nowrap',
    textDecoration: 'none',
    borderTop: '1px solid transparent',
    borderBottom: (props: Props) =>
      `1px solid ${
        props.isActive && props.placement === 'topbar'
          ? theme.palette.primary.main
          : 'transparent'
      }`,
    '-webkit-tap-highlight-color': 'transparent',

    '&:hover': {
      color: theme.palette.navigationLink.activeColor,
      fill: theme.palette.navigationLink.activeColor
    },

    [theme.breakpoints.down('lg')]: {
      paddingLeft: theme.spacing(4),
      borderBottom: (props: Props) =>
        props.isActive && props.placement !== 'mobileDropdown'
          ? `1px solid ${theme.palette.primary.main}`
          : 'none'
    }
  },
  text: {
    color: 'inherit'
  },
  icon: {
    fill: 'inherit',
    '& path': {
      fill: 'inherit'
    }
  }
}))
