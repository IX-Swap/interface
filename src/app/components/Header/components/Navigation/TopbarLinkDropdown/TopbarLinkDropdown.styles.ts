import makeStyles from '@mui/styles/makeStyles'

const importantPosition = '0!important'

export const useStyles = makeStyles(theme => {
  const lineWidth = `calc(100% - 32px)`

  return {
    wrapper: {
      display: 'flex'
    },
    paper: {
      marginTop: importantPosition,
      padding: importantPosition,
      borderTopRightRadius: importantPosition,
      borderTopLeftRadius: importantPosition,
      top: '80px!important',
      boxShadow: `0px 80px 80px ${theme.palette.dropdownLink.boxShadow}`,
      border: `1px solid ${theme.palette.dropdownLink.border}!important`
    },
    list: { padding: 0 },
    navItem: {
      position: 'relative',
      minWidth: 266,
      height: 'max-content',
      padding: theme.spacing(0),

      '&:hover': {
        background: 'initial'
      }
    },
    line: {
      position: 'absolute',
      bottom: 0,
      height: 1,
      width: lineWidth,
      backgroundColor: theme.palette.dropdownLink.border,
      marginRight: theme.spacing(2),
      marginLeft: theme.spacing(2)
    }
  }
})
