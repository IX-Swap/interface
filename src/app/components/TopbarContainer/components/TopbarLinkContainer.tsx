import React from 'react'
import { useLocation } from 'react-router-dom'
import { useAppBreakpoints } from 'hooks/useAppBreakpoints'
import { HeaderNavigationLink } from 'app/components/TopbarContainer/components/HeaderNavigationLink'
import { blue } from '@material-ui/core/colors'

export interface TopbarLinkProps {
  link: string
  label: string
  disabled?: boolean
  colorVariant?: 'nav' | 'menu'
}

export const TopbarLinkContainer = (props: TopbarLinkProps) => {
  const { link, label, disabled = false, colorVariant = 'nav' } = props
  const { pathname } = useLocation()

  const baseLink = link.split('/').slice(0, 3).join('/')
  const isActive = pathname.startsWith(baseLink)

  const { theme } = useAppBreakpoints()
  const navColor = isActive
    ? blue[300]
    : theme.palette.type === 'dark'
    ? theme.palette.getContrastText(theme.palette.background.default)
    : theme.palette.background.default
  const menuColor = isActive
    ? theme.palette.background.default
    : theme.palette.slider.activeBackground

  return (
    <HeaderNavigationLink
      to={link}
      style={{
        color: colorVariant === 'menu' ? menuColor : navColor
      }}
      onClick={e => (disabled ? e.preventDefault() : null)}
    >
      <span style={{ whiteSpace: 'nowrap' }}>{label}</span>
      {disabled ? (
        <svg
          width='7'
          height='4'
          viewBox='0 0 7 4'
          fill='none'
          xmlns='http://www.w3.org/2000/svg'
        >
          <path
            d='M3.5 4L0.468911 0.25L6.53109 0.25L3.5 4Z'
            fill={
              colorVariant === 'nav' && isActive
                ? blue[300]
                : theme.palette.type === 'dark'
                ? theme.palette.getContrastText(
                    theme.palette.background.default
                  )
                : theme.palette.background.default
            }
          />
        </svg>
      ) : null}
    </HeaderNavigationLink>
  )
}
