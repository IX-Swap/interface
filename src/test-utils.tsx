import React from 'react'
import { Router } from 'react-router-dom'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import {
  createGenerateClassName,
  StylesProvider
} from '@material-ui/core/styles'
import { ThemeProvider } from '@material-ui/styles'
import Themes from './v2/themes'
import { history } from 'v2/history'
import { UserProvider } from 'v2/auth/context'
import { UserStore } from 'v2/auth/context/store'
import { PasswordResetProvider } from 'v2/auth/context/password-reset'
import PasswordResetStore from 'v2/auth/context/password-reset/store'
import { PasswordResetStep } from 'v2/auth/context/password-reset/types'
import { AuthorizerTableStore } from 'v2/app/pages/authorizer/context/store'
import { AuthorizerTableStoreProvider } from 'v2/app/pages/authorizer/context'
import { DepositStore } from 'v2/app/pages/accounts/pages/banks/context/store'
import { DepositStoreProvider } from 'v2/app/pages/accounts/pages/banks/context'
import { ServicesProvider } from 'v2/services/useServices'
import { renderHook, RenderHookResult } from '@testing-library/react-hooks'
import { BreadcrumbsProvider } from 'v2/hooks/useBreadcrumbs'

const generateClassName = createGenerateClassName({
  productionPrefix: 'ix'
})

const BaseProviders: React.FC = ({ children }) => {
  return (
    <StylesProvider generateClassName={generateClassName}>
      <ThemeProvider theme={Themes.default}>
        <BreadcrumbsProvider>
          <Router history={history}>{children}</Router>
        </BreadcrumbsProvider>
      </ThemeProvider>
    </StylesProvider>
  )
}

const customRenderer = (
  ui: any,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => {
  return render(ui, { wrapper: BaseProviders, ...options })
}

export const testComponentId = 'test-component'

export const fakeUserStore: Partial<UserStore> = {
  activeTab: 0,
  setActiveTab: jest.fn()
}

export const renderWithUserStore = (
  ui: any,
  store?: Partial<UserStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <UserProvider value={{ ...fakeUserStore, ...store }}>
        {children}
      </UserProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakePasswordResetStore: Partial<PasswordResetStore> = {
  currentStep: PasswordResetStep.Request,
  setCurrentStep: jest.fn()
}

export const renderWithPasswordResetStore = (
  ui: any,
  store?: Partial<PasswordResetStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <PasswordResetProvider value={{ ...fakePasswordResetStore, ...store }}>
        {children}
      </PasswordResetProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakeAuthorizerTableStore: Partial<AuthorizerTableStore> = {
  idKey: '__',
  uri: '/api/source',
  _getItemId: jest.fn()
}

export const renderWithAuthorizerTableStore = (
  ui: any,
  store?: Partial<AuthorizerTableStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <AuthorizerTableStoreProvider
        value={{ ...fakeAuthorizerTableStore, ...store }}
      >
        {children}
      </AuthorizerTableStoreProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakeDepositStore: Partial<DepositStore> = {}

export const renderWithDepositStore = (
  ui: any,
  store?: Partial<DepositStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <DepositStoreProvider value={{ ...fakeDepositStore, ...store }}>
        {children}
      </DepositStoreProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const renderHookWithServiceProvider = (
  hookFn: any,
  store: object = {}
): RenderHookResult<any, any> => {
  const WithServiceProvider: React.FC = ({ children }) => (
    <ServicesProvider value={store}>{children}</ServicesProvider>
  )

  return renderHook(hookFn, { wrapper: WithServiceProvider })
}

export * from '@testing-library/react'
export { customRenderer as render }
