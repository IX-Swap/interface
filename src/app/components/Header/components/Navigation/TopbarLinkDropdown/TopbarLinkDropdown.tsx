import React from 'react'
import { Box, Menu, MenuItem } from '@mui/material'
import { InternalRouteProps } from 'types/util'
import { TopbarLinkContainer } from 'app/components/Header/components/Navigation/TopbarLinkContainer/TopbarLinkContainer'
import { useStyles } from 'app/components/Header/components/Navigation/TopbarLinkDropdown/TopbarLinkDropdown.styles'

export interface TopbarLinkDropdownProps {
  link: string
  label: string
  linkItems: InternalRouteProps[]
}

export const TopbarLinkDropdown = ({
  link,
  label,
  linkItems
}: TopbarLinkDropdownProps) => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null)
  const classes = useStyles()

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <>
      <div className={classes.wrapper} onClick={e => handleClick(e)}>
        <TopbarLinkContainer
          placement={'topbar'}
          link={link}
          label={label}
          disabled={true}
          active={Boolean(anchorEl)}
        />
      </div>

      <Menu
        classes={{
          paper: classes.paper,
          list: classes.list
        }}
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        {linkItems.map((accountLink, i) => {
          return (
            <MenuItem
              onClick={handleClose}
              key={accountLink.label}
              className={classes.navItem}
              disableTouchRipple
            >
              <TopbarLinkContainer
                label={accountLink.label}
                link={accountLink.path}
                key={accountLink.label}
              />
              {i < linkItems.length - 1 && <Box className={classes.line} />}
            </MenuItem>
          )
        })}
      </Menu>
    </>
  )
}
