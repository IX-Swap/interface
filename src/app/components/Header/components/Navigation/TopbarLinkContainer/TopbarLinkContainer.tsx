import React from 'react'
import { useLocation, Link } from 'react-router-dom'
import ArrowDropDownIcon from '@mui/icons-material/ArrowDropDown'
import { Typography } from '@mui/material'
import { useStyles } from 'app/components/Header/components/Navigation/TopbarLinkContainer/TopbarLinkContainer.styles'
import { TwoFADialogWrapper } from 'app/components/TwoFADialogWrapper'
import { IssuanceRoute } from 'app/pages/issuance/router/config'
import { useUncompletedIdentityDialog } from 'app/components/UncompletedIdentityDialog/hook/useUncompletedIdentityDialog'
import { useAppNavigation } from 'app/components/Header/hooks/useAppNavigation'

export interface TopbarLinkProps {
  link: string
  label: string
  disabled?: boolean
  placement?: 'topbar' | 'dropdown' | 'mobileDropdown'
  active?: boolean
  isExternalUrl?: boolean
  onClick?: () => void
}

export const TopbarLinkContainer = (props: TopbarLinkProps) => {
  const {
    link,
    label,
    disabled = false,
    placement = 'dropdown',
    active = false,
    isExternalUrl = false,
    onClick
  } = props
  const { pathname } = useLocation()
  const baseLink = link.split('/').slice(0, 3).join('/')

  const { isNavigationImpossibleWithoutCompletedIdentity } = useAppNavigation()
  const { showUncompletedIdentityDialog } = useUncompletedIdentityDialog(link)

  const getIsLinkActive = () => {
    if (placement === 'topbar') {
      return pathname.startsWith(baseLink)
    }

    // This check is needed because the pathname that contains the 'exchange' also contains pairId
    if (link.includes('exchange') && pathname.includes('exchange')) {
      return true
    }

    return pathname === link
  }
  const isActive = getIsLinkActive()

  const classes = useStyles({
    isActive: active || isActive,
    disabled,
    placement
  })

  const handleClick = (e: React.MouseEvent) => {
    if (disabled || isNavigationImpossibleWithoutCompletedIdentity(link)) {
      e.preventDefault()
      !disabled && showUncompletedIdentityDialog()
    }

    if (onClick !== undefined) {
      onClick()
    }
  }

  return (
    <TwoFADialogWrapper
      extraCheck={
        link === IssuanceRoute.create || link === IssuanceRoute.createListing
      }
    >
      <Link
        to={!isExternalUrl ? link : { pathname: link }}
        className={classes.wrapper}
        target={!isExternalUrl ? '' : '_blank'}
        onClick={handleClick}
      >
        <Typography className={classes.text} variant={'body1'}>
          {label}
        </Typography>
        {disabled ? <ArrowDropDownIcon className={classes.icon} /> : null}
      </Link>
    </TwoFADialogWrapper>
  )
}
