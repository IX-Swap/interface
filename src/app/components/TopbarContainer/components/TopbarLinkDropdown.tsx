import React from 'react'
import { Menu } from '@material-ui/core'
import { TopbarLinkContainer } from 'app/components/TopbarContainer/components/TopbarLinkContainer'
import { InternalRouteProps } from 'types/util'
import { TopNavMenuItem } from 'app/components/TopbarContainer/components/TopNavMenuItem'

export interface TopbarLinkDropdownProps {
  link: string
  label: string
  linkItems: InternalRouteProps[]
}

export const TopbarLinkDropdown = ({
  link,
  label,
  linkItems
}: TopbarLinkDropdownProps) => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null)

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <>
      <div style={{ display: 'flex' }} onClick={e => handleClick(e)}>
        <TopbarLinkContainer link={link} label={label} disabled={true} />
      </div>

      <Menu
        style={{ top: 72, paddingTop: 0, paddingBottom: 0 }}
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        {linkItems.map(accountLink => {
          return (
            <TopNavMenuItem onClick={handleClose}>
              <TopbarLinkContainer
                label={accountLink.label}
                link={accountLink.path}
                key={accountLink.label}
                colorVariant={'menu'}
              />
            </TopNavMenuItem>
          )
        })}
      </Menu>
    </>
  )
}
