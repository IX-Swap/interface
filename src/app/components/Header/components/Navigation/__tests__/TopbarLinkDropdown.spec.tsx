import React from 'react'
import { render } from 'test-utils'
import { TopbarLinkDropdown } from 'app/components/Header/components/Navigation/TopbarLinkDropdown/TopbarLinkDropdown'
import { ExternalNavigationLink } from 'app/components/Header/components/Navigation/TopbarLinkContainer/ExternalNavigationLink'

const props = {
  link: '/test',
  label: 'label',
  linkItems: []
}
jest.mock(
  'app/components/Header/components/Navigation/TopbarLinkContainer/ExternalNavigationLink',
  () => ({
    ExternalNavigationLink: jest.fn(() => null)
  })
)
const externalProps = {
  ...props,
  linkItems: [{ label: 't', path: '/test', external: true }]
}

describe('TopbarLinkDropdown', () => {
  afterEach(async () => {
    jest.clearAllMocks()
  })

  it('should match snapshot', () => {
    const { container } = render(<TopbarLinkDropdown {...props} />)

    expect(container).toMatchSnapshot()
  })

  it('should render external link', () => {
    render(<TopbarLinkDropdown {...externalProps} />)

    expect(ExternalNavigationLink).toHaveBeenCalledWith(
      expect.objectContaining({
        link: externalProps.linkItems[0].path,
        label: externalProps.linkItems[0].label
      }),
      {}
    )
  })
})
