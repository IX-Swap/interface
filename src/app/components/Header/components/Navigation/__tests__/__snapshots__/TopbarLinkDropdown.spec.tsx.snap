// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopbarLinkDropdown should match snapshot 1`] = `
<div>
  <div
    class="makeStyles-wrapper-1"
  >
    <a
      class="makeStyles-wrapper-6 makeStyles-wrapper-9"
      href="/test"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 makeStyles-text-7 css-1b4l2pg-MuiTypography-root"
      >
        label
      </p>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium makeStyles-icon-8 css-1oaawcc-MuiSvgIcon-root"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m7 10 5 5 5-5z"
        />
      </svg>
    </a>
  </div>
</div>
`;
